File: message_passing.py


import io, struct, select, time
from PIL import Image

""" This library is used to send simple messages to and
	from the microcontroller. Features include:
	-thermostat control
	-touchscreen control
	-image streaming
"""

# Serial1 connection
__SERIAL1 = '/dev/ttyATH0'

# Timeout and maximum number of trials
__BAUD = 9600 # will increase this in the future
__TIMEOUT = 1 # seconds
__MAX_TRIALS = 3

# Message commands constants
__GET_ROOM_TEMP 		= 'R'
__GET_TARGET_TEMP 		= 'G'
__GET_MODE 				= 'H'
__GET_UNIT				= 'J'
__IS_ON					= 'O'
__SET_TARGET_TEMP 		= 'S'
#__INCREMENT_TARGET_TEMP = 'I'
#__DECREMENT_TARGET_TEMP = 'D'
__SWITCH_MODE			= 'M'
__SWITCH_UNIT			= 'U'
__WRITE_TEXT			= 'W'
__STREAM_IMAGE			= 'P'

# Image streaming sub-commands
__STREAM_INIT			= 'i'
__STREAM_SEND			= 's'
__STREAM_FIN			= 'f'

# Gets the temperature of the room
def getRoomTemp():
	result = __sendCommandWithRetry(__GET_ROOM_TEMP, '', True)
	if(result.isdigit()):
		return int(result)
	else:
		return 'error'

# Gets the target temperature
def getTargetTemp():
	result = __sendCommandWithRetry(__GET_TARGET_TEMP, '', True)
	if(result.isdigit()):
		return int(result)
	else:
		return 'error'

# Gets the current mode of operation (heating or cooling)
def getMode():
	result = __sendCommandWithRetry(__GET_MODE, '', True)
	if(result == 'heating' or result == 'cooling'):
		return result
	else:
		return 'error'

# Gets the temperature units (f or c)
def getUnit():
	result = __sendCommandWithRetry(__GET_UNIT, '', True)
	if(result == 'f' or result == 'c'):
		return result
	else:
		return 'error'

# Return True if the system is on, False otherwise
def isOn():
	result = __sendCommandWithRetry(__IS_ON, '', True)
	return (result == 'true')

# Sets the target temperature
def setTargetTemp(newTargetTemp):
	# tempString = '%d' % newTargetTemp
	temp = struct.pack('h', newTargetTemp)
	return __sendCommandWithRetry(__SET_TARGET_TEMP, temp, False)

# # Increase the target temperature by one degree
# def incrementTargetTemp():
# 	return __sendCommandWithRetry(__INCREMENT_TARGET_TEMP, '', False)

# # Decrease the target temperature by one degree
# def decrementTargetTemp():
# 	return __sendCommandWithRetry(__DECREMENT_TARGET_TEMP, '', False)

# Switches the system between heating and cooling
def switchMode():
	return __sendCommandWithRetry(__SWITCH_MODE, '', False)

# Switches the temperature units
def switchUnit():
	return __sendCommandWithRetry(__SWITCH_UNIT, '', False)

# Sends text to be displayed on the screen
def writeText(x, y, fColor, bColor, fontSize, text):
	packet = '%s%s' % (struct.pack('HHHHB', x, y, fColor, bColor, fontSize), text)
	return __sendCommandWithRetry(__WRITE_TEXT, packet, False)

# Sends an image touchscreen at the specified location
def streamImage(file, xpos, ypos):
	image = Image.open(file)
	width, height = image.size
	# prep image for sending
	r, g, b = image.convert('RGB').split()
	R = list(r.getdata())
	G = list(g.getdata())
	B = list(b.getdata())
	imagePixels = R
	for i in range(len(R)):
		imagePixels[i] = ((R[i] & 0xF8) << 8) | ((G[i] & 0xFC) << 3) | (B[i] >> 3)
	# send an image command
	message = struct.pack('cHHHH', __STREAM_INIT, xpos, ypos, width, height)
	response = __sendCommandWithRetry(__STREAM_IMAGE, message, True)
	print response
	if response == 'error' or len(response) != 4:
		return 'error'
	else:
		bufferSize, firstPixel = struct.unpack('HH', response)
		firstPixel *= 8
	# send image file
	if __stream(imagePixels, firstPixel, bufferSize) == 'error':
		return 'error'
	# send termination command
	__sendCommandWithoutRetry(__STREAM_IMAGE, __STREAM_FIN)

# Sends a command to the microcontroller with the message
# expectedResponse should be True if expecting a response
# from the microcontroller, False otherwise
def __sendCommandWithRetry(command, message, expectedResponse):
	packet = '%s%s\n' % (command, message)
	# transmit data and wait for response
	trial = 0
	success = False
	# TODO add in handshake so no duplicate packets?
	while (not success) and (trial < __MAX_TRIALS):
		ser_out = io.open(__SERIAL1, 'wb')
		ser_out.write(packet)
		ser_out.close()
		ser_in = io.open(__SERIAL1, 'rb')
		rlist, wlist, xlist = select.select([ser_in], [], [], __TIMEOUT)
		for reader in rlist:
			response = reader.readline().rstrip('\r\n')
			# print response
			success = (len(response) > 0) and (response[0] == command)
		trial += 1
		ser_in.close()

	if success:
		if expectedResponse:
			return response[1:]
	else:
		return 'error'

# Sends a command to the microcontroller with the message
# Does not wait for any response before terminating only sends the
# packet once
def __sendCommandWithoutRetry(command, message):
	packet = '%s%s\n' % (command, message)
	ser_out = io.open(__SERIAL1, 'wb')
	ser_out.write(packet)
	ser_out.close()

# Sends an image to the microcontroller using an ACK-based protocol
# May error out if the microcontroller does not respond
def __stream(data, firstPixel, bufferSize):
	headerSize = 6 # bytes
	pixelsPerTx = int((bufferSize - headerSize) / (2 * 8)) * 8 # get a multiple of 8 pixels
	pixelExpected = firstPixel # pixels
	totalPixels = len(data) # pixels
	trial = 0
	print pixelExpected, totalPixels
	while (pixelExpected < totalPixels) and (trial < __MAX_TRIALS):
		print 'sending pixel %d' % pixelExpected

		success = False
		pixelsToTx = min(int((totalPixels - pixelExpected) / 8) * 8, pixelsPerTx)
		# write to fill up the buffer or end of file
		ser_out = io.open(__SERIAL1, 'wb')
		ser_out.write(struct.pack('ccHH', __STREAM_IMAGE, __STREAM_SEND, pixelExpected / 8, pixelsToTx / 8))
		for i in range(pixelExpected, pixelsToTx + pixelExpected):
			# print hex(data[i])
			ser_out.write(struct.pack('H', data[i])) #TODO pad the end with zeros if needed
		ser_out.close()
		time.sleep(0.1)
		# read back response
		ser_in = io.open(__SERIAL1, 'rb')
		rlist, wlist, xlist = select.select([ser_in], [], [], __TIMEOUT)
		for reader in rlist:
			response = reader.readline().rstrip('\r\n')
			if (len(response) == 3) and (response[0] == __STREAM_IMAGE):
				pixelExpected, = struct.unpack('H', response[1:3])
				pixelExpected *= 8
				success = True
		if success:
			trial = 0
		else:
			trial += 1
		ser_in.close()
	if(trial == __MAX_TRIALS):
		return 'error'

# Sends an image to the microcontroller with little to no checking
def __stream2(data, bufferSize):
	pixelsPerTx = int((bufferSize - 2) / 2)	
	nextPixel = 0
	while(nextPixel < len(data)):
		ser_out = io.open(__SERIAL1, 'wb')
		ser_out.write(struct.pack('cc', __STREAM_IMAGE, __STREAM_SEND))
		for i in range(nextPixel, min(nextPixel + pixelsPerTx, len(data))):
			ser_out.write(struct.pack('H', data[i]))
		ser_out.close()
		ser_in = io.open(__SERIAL1, 'rb')
		rlist, wlist, xlist = select.select([ser_in], [], [], 3) #longer timeout
		for reader in rlist:
			response = reader.readline().rstrip('\r\n')
			if (len(response) > 0) and (response[0] == __STREAM_IMAGE):
				nextPixel += pixelsPerTx
		ser_in.close()
		
#Used for testing only
def testStream(data_size):
	data = []
	for i in range(data_size):
		data.append(i+1)
	# send an image command
	print 'sending init...'
	message = struct.pack('cHHHH', __STREAM_INIT, 32, 32, 8, len(data) / 8)
	response = __sendCommandWithRetry(__STREAM_IMAGE, message, True)
	#print len(response)
	if response == 'error' or len(response) != 4:
		return 'error'
	else:
		bufferSize, firstPixel = struct.unpack('HH', response)
		# print firstPixel, bufferSize
		firstPixel *= 8

	# send image file
	print 'sending image...'
	# if __stream(data, firstPixel, bufferSize) == 'error':
	# 	return 'error'
	__stream2(data, bufferSize)

	# send termination command
	#print 'sending temination...'
	#__sendCommandWithoutRetry(__STREAM_IMAGE, __STREAM_FIN)

File: messagePassingTest/MessagePassing.cpp


#include "MessagePassing.h"

// Interprets commands from Linino for controlling the thermostat and touchscreen
void processCommands(TempControl *tc) {
  char ret[COMMAND_BUFFER_SIZE];
//  Serial.println("processing...");
//  Serial.println(Serial1.available());
  while(Serial1.available() > 0) {
    memset(ret, 0x00, sizeof(ret));
    int numBytes = Serial1.readBytesUntil('\n', ret, COMMAND_BUFFER_SIZE);
    Serial.println("number of bytes read: " + String(numBytes));
    switch(ret[0]) {
      case GET_ROOM_TEMP:
        Serial.println("get room temperature");
        Serial1.println(String(GET_ROOM_TEMP) + String(tc->getRoomTemp()));
        break;
      case GET_TARGET_TEMP:
        Serial.println("get target temperature");
        Serial1.println(String(GET_TARGET_TEMP) + String(tc->getTargetTemp()));
        break;
      case GET_MODE:
        Serial.println("get mode");
        Serial1.println(String(GET_MODE) + modeToString(tc));
        break;
      case GET_UNIT:
        Serial.println("get unit");
        Serial1.println(String(GET_UNIT) + unitToString(tc));
        break;
      case IS_ON:
        Serial.println("is on?");
        Serial1.println(String(IS_ON) + isOnToString(tc));
        break;
      case SET_TARGET_TEMP:
        Serial.println("set target temperature");
        tc->setTargetTemp(interpretNumber(ret, sizeof(ret)));
        Serial1.println(String(SET_TARGET_TEMP));
        break;
      case INCREMENT_TARGET_TEMP:
        Serial.println("increment target temperature");
        tc->incrementTargetTemp();
        Serial1.println(String(INCREMENT_TARGET_TEMP));
        break;
      case DECREMENT_TARGET_TEMP:
        Serial.println("decrement target temperature");
        tc->decrementTargetTemp();
        Serial1.println(String(DECREMENT_TARGET_TEMP));
        break;
      case SWITCH_MODE:
        Serial.println("switch mode");
        tc->switchMode();
        Serial1.println(String(SWITCH_MODE));
        break;
      case SWITCH_UNIT:
        Serial.println("switch unit");
        tc->switchUnit();
        Serial1.println(String(SWITCH_UNIT));
        break;
      case STREAM_IMAGE:
        Serial.println("stream image");
        imageInitResponse();
        receiveImage(unpackNumber(ret, 1, 2), unpackNumber(ret, 3, 2),
                      unpackNumber(ret, 5, 2), unpackNumber(ret, 7, 2));
        break;
      default:
        Serial.println("Error: message command not recognized");
        // Serial.println(ret);
    }
  }
}

// Returns a String representation of the thermostat's mode 
String modeToString(TempControl *tc) {
  switch(tc->getMode()) {
    case HEATING:
      return "heating";
    case CELCIUS:
      return "cooling";
  }
  return "";
}

// Returns a String representation of the current units of the thermostat
String unitToString(TempControl *tc) {
  switch(tc->getUnit()) {
    case FAHRENHEIT:
      return "f";
    case CELCIUS:
      return "c";
  }
  return "";
}

// Returns a String representation of the whether the thermostat is on or off
String isOnToString(TempControl *tc) {
  if(tc->isOn()) {
    return "true";
  } else {
    return "false";
  }
}

int interpretNumber(char *number, int len) { //TODO to be removed and changed to binary
  int output = 0;
  int i;
  for(i = 1; i < len && isdigit(number[i]); i++) {
    output *= 10;
    output += number[i] - '0';
  }
  return output;
}

// Grabs the number representation of the a sequence of bytes
uint16_t unpackNumber(char *bytes, int start, int len) {
  uint16_t output = 0;
  int i;
  for(i = start; i < start + len; i++) {
    output = output << 8;
    output = output | (0x00FF & bytes[i]);
  }
  return output;
}

// Processes the image and forwards the data to the screen
void receiveImage(uint16_t xpos, uint16_t ypos, uint16_t width, uint16_t height) {
  // allocate the buffer and start receiving data
  char image[IMAGE_BUFFER_SIZE];
  uint8_t failCount = 0;
  bool success = false;
  uint16_t pixelCount = 0;  // grouped in 8s
  while((failCount < MAX_FAIL_COUNT) && !success) {
    // read next packet
    Serial.println("in the loop");
    memset(image, 0x00, sizeof(image));
    uint16_t bytesRead = Serial1.readBytes(image, sizeof(image));
    int i;
    Serial.println("read bytes: " + String(bytesRead));
    for(i = 0; i < bytesRead; i++) { //TODO remove this loop, only prints packet contents
      if(image[i] == 0) {
        Serial.print("00");
      } else {
        Serial.print((0x00FF & image[i]), HEX);
      }
    }
    Serial.println();
    // interpret packet
    if(image[0] == STREAM_IMAGE) {
      switch(image[1]) {
        case STREAM_INIT:
          Serial.println("init");
          imageInitResponse();
          failCount = 0;
          break;
        case STREAM_SEND:
          Serial.println("send");
          pixelCount += imageSendResponse(image, bytesRead, pixelCount);
          failCount = 0;
          break;
        case STREAM_FIN:
          Serial.println("finished");
          success = true;
          break;
        default:
          failCount++;
      }
    } else {
      failCount++;
    }
  }
}

// Sends an init response to begin image streaming
void imageInitResponse() {
  char responseInit[5] = {STREAM_IMAGE, (IMAGE_BUFFER_SIZE & 0xFF00) >> 8, (IMAGE_BUFFER_SIZE & 0x00FF), 0, 0};
  writeResponse(responseInit, sizeof(responseInit));
}

// Sends a send response for image streaming
uint16_t imageSendResponse(char *image, int len, uint16_t pixelCount) {
  char responseSend[3] = {STREAM_IMAGE, 0, 0};
  uint16_t pixelsRead = 0; // grouped in 8s
  int byteCount = 0;
  bool finished = false;
  while(!finished && (byteCount < len) && (image[byteCount] == STREAM_IMAGE) && (image[byteCount + 1] == STREAM_SEND)) {
    int pixelsTemp = imageToScreen(image + byteCount, len - byteCount, pixelCount + pixelsRead);
    if(pixelsTemp != 0) {
      pixelsRead += pixelsTemp;
      byteCount += pixelsTemp * 16 + 6;
    } else {
      finished = true;
    }
  }
  responseSend[1] = (0xFF00 & (pixelCount + pixelsRead) >> 8;
  responseSend[2] = (0x00FF & (pixelCount + pixelsRead));
  writeResponse(responseSend, sizeof(responseSend));
  return pixelsRead;
}

// Sends a response to the Linino on Serial1
void writeResponse(char *response, int len) {
  int i;
  for(i = 0; i < len; i++) {
    Serial1.print(response[i]);
  }
  Serial1.println();
}

// Processes a single packet and sends pixel data to the screen
// Returns the number of pixels actually written
uint16_t imageToScreen(char *image, int len, uint16_t pixelCount) {
  uint16_t firstPixel = unpackNumber(image, 2, 2);
  uint16_t pixelsInPack = unpackNumber(image, 4, 2);
//  Serial.println(String(pixelsInPack) + ", " + String(bytesRead));
//  Serial.println(String(pixelsInPack * 16 + 6));
  if((pixelsInPack * 16 + 6 > len) || (firstPixel != pixelCount)) { // TODO grab the end of the packet if necessary
    return 0;
  } else {
    return pixelsInPack;
  }
}


